### ISSUE TITLE
Execute M2 spellbook ranker on movement-active world (DEC-042)

### ISSUE THREAD

**Issue by gamesh411** [gamesh411/mod-playerbots#28]
Part of #4.

## Question

Execute the M2 ability-head co-adaptation per **DEC-042** (design locked in #24):

- Land the `duel_v6` CSV rev once #25/#26 fix the final vector/vocab.
- BC bootstrap on `expert_action` with the 128-vs-256 hidden sweep (winner by holdout CE + teacher agreement; record it here).
- DEC-033 explore + win-anchored loop, no teacher after round 0; degeneracy gate per deploy; anti-thrash budget 2.
- Farm: frozen M1 canonical movement heads both seats, movement tau=10; abilities explore.
- Gate: both-seat WR uplift delta 0.02 vs fresh S0-sentinel+M1 baseline (epoch 2, DEC-037 rules, mixed-seat, >=2.4k matches/run, tau=0 both channels); dec033r3+M1 and dec033r3+M0 report-only showcase rows.
- Freeze per DEC-019: `stage/m2-<slug>`, `artifacts/duel/m2/`, manifest, stage card, `duel-m2` replay profile row (DEC-040 extension), data tag `duel_v6`.

Blocked by: #25, #26 (feature/vocab decisions land before the farm spins).

**Comment by gamesh411** [comment]
## Round-2 farm: overnight OS sleep, then restarted for throughput (CSV rotated)

Two events since launch, both survived cleanly:

**1. The machine slept 01:33-08:42** (Kernel-Power 42 / Power-Troubleshooter 1). The worldserver froze with it and resumed on wake; the duel timeline shows a clean 6.5 h silence with zero duels started inside it, zero malformed rows, and only the handful of straddling duels carrying odd wall-clock durations - win-anchored labels are per-duel outcomes, so unaffected. Lesson for farm nights: disable AC sleep first (`powercfg /change standby-timeout-ac 0`).

**2. Restarted at 13:33** when throughput sagged to ~5.2k duels/h (round-1 playbook threshold ~8k/h). Same `match_id`-collision discipline as round 1 - rotated, never appended:

- boot1 (frozen): `ml_decisions_duel_m2_r2_boot1.csv` - **3,063,032 rows / 50,204 duels**, 0 wrong-width
- boot2 (live): `ml_decisions_duel_m2_r2.csv` - fresh 135-col duel_v6 header

All four heads loaded explicitly on the new boot (m2 r1-win in=112 both seats, M1 movement in=90); conf unchanged (`-SkipPrereqs -SkipBuild -SkipExtract`, no conf re-apply needed since the log knob names the same file).

Combined volume **~3.08M rows (77% of ~4M)**; remaining ~940k is ~1.5-2 h at the 13k/h opening rate. Train with `--csv ..._boot1.csv ..._m2_r2.csv` so outcomes stay per-file.

**Comment by gamesh411** [comment]
## Round-2 win-anchored heads trained; both deploy gates PASS; round 3 wired (not started)

Farm stopped at 18:27 with **5,903,390 rows / 95,902 duels** across the two boots (148% of the ~4M target - the overnight-sleep session ran long after the 13:33 rotation). Train recipe unchanged from round 1: `--label-scheme win-only --balance-labels sqrt --hidden 256 --epochs 40`, frozen r0 vocab, both CSVs passed per-file.

| | train rows | won_frac | holdout CE (r1 -> r2) | distinct argmax (r1 -> r2) | gate @ 20k |
|---|---|---|---|---|---|
| warrior | 1,688,193 | 0.718 | 2.62 -> **2.57** | 37/45 -> **39/45** | 36/45 |
| mage | 427,542 | 0.242 | 4.61 -> **4.71** | 20/218 -> **103/218** | 81/218 |

### Gates (boot1 holdout, `match_id % 10 == 7`)

Warrior: PASS. Battle Stance 23.3%, Berserker Stance 15.8%, Pummel 7.6%, Bloodrage 7.3%, Defensive Stance 5.4%, Sweeping Strikes, Recklessness, Charge, Disarm, Intercept, Rend, Heroic Strike. Expert-agreement 27.6%. No sink.

Mage: PASS. Icy Veins 27.2%, Counterspell 20.0%, Cold Snap 7.3%, **Summon Water Elemental 6.0%**, Freeze, Deep Freeze, Ice Block kit. Expert-agreement 4.9%.

```
pet-down floor: spell 31687 is argmax at some point in 43.0% of 5020
  pet-down duel-seats (floor 25%) -> PASS
  report-only: argmax at the opening ready tick in 0.0%
```

(Round 1 was 51.2% ever / 0.0% opening; the learned deferral holds.)

### The asymmetry worth watching

won_frac moved 0.624/0.347 (r1 data) -> **0.718/0.242**: the round-1 warrior dominated the round-2 world. That thins the mage's win-only training set (428k rows vs the warrior's 1.69M) and its CE ticked up while its argmax spread exploded (20 -> 103 distinct). The spread explosion reads as the win-only signal fragmenting over rarer won-duel patterns rather than degeneracy - the gate's 81 distinct argmax with a coherent Frost top-kit supports that - but if round 3 pushes won_frac further apart, the mage seat may be starving and the loop's plateau check should weigh both seats' win rates, not just uplift.

### Deployed + wired

- Module `d6c5b180`: canonical `*.pbml` <- `.r2-win.pbml` (sha256-verified), snapshots committed.
- Orchestrator `ed31422`: log rotates to `ml_decisions_duel_m2_r3.csv`, recipe otherwise unchanged.
- **Round-3 farm is wired, not launched** - starting it stays a separate call.

Anti-thrash budget still intact at 2 (no round rerun yet).

**Comment by gamesh411** [comment]
## Round-2 gate measured: warrior parity, mage -18.8pp. First WR number of the M2 loop.

Full DEC-042 protocol - mixed seat, tau=0 both channels, delta 0.02, DEC-037 rules, fresh S0-sentinel+M1 baseline. Every leg cleared the >=2.4k minimum comfortably (450 bots, 96.4% duel saturation, 210 concurrent matches).

| Run | Seat | WR | Matches | Threshold | Verdict |
|---|---|---|---|---|---|
| Baseline: S0 sentinel + M1 movement | warrior | 45.9% | 10,709 | - | - |
| Baseline: S0 sentinel + M1 movement | mage | 54.6% | 10,645 | - | - |
| r2 **arms-ranker** | warrior | **46.9%** | 5,308 | >=47.9% | **FAIL** - +1.0pp, ~1.2 sigma (parity) |
| r2 **frost-ranker** | mage | **35.8%** | 7,875 | >=56.6% | **FAIL** - -18.8pp, ~26 sigma |

CSVs: `ml_decisions_duel_m2gate_{base,arms,frost}.csv`. Anti-thrash budget untouched (this is a measurement, not a round rerun).

The baseline is fresh per DEC-042 and is **not** comparable to M1's 39.3/61.5 M0<->M0 figure - that one predates DEC-044 (pet reset on duel end, 25% unglyphed mages). The 39.3 -> 45.9 and 61.5 -> 54.6 shift is that engineering, measured.

### This is the starvation worry from the round-2 comment, confirmed

The previous comment flagged won_frac 0.718/0.242 as possible mage starvation and asked that the plateau check weigh both seats' win rates rather than uplift alone. It does now, and the mage seat is far below stock rather than merely flat. Warrior lands at parity - the same shape M1 hit in DEC-041, where a near-optimal scripted opponent leaves no headroom.

The head is **not** degenerate. At tau=0 it summons the elemental in 32.7% of duels and shows 53 distinct argmax on gate states. It is simply a weaker policy than the stock rotation: 12.4% of its decisions go to wand shoot and 7.3% to Counterspell, against a baseline rotation of 21.1% frostbolt / 8.8% frost nova / 8.0% deep freeze.

Worth recording as a loop-discipline finding: **the per-deploy degeneracy gate cannot see this.** It asks "is this action dead", never "is this policy good", and both round-2 heads passed it. Related trap - the DEC-050 pet-down floor reads **43.0%** on tau=10 farm states and **0.1%** on tau=0 gate states for the identical head. It is a farm-distribution metric and must not be read on gate data.

### Measured on the post-upstream-sync stack

This ran on the rebased core+modules (`pre/post-upstream-sync-20260816`). The gate numbers are curriculum, not sync damage:

- `eval_duel_winrate.py` reproduces M1's frozen baseline exactly from its archived CSV: **39.3% / 61.4%** vs the manifest's 39.3% / 61.5%.
- A clone movement head scores **0.9828** teacher agreement on post-sync data - feature vector aligned.
- The M2 warrior head measures at parity; a broken vector would not spare one seat.
- DEC-050 pet-down floor at matched tau=10: **43.0%** pre-sync vs **39.7%** post-sync.
- Baseline stock rotation is textbook Frost (frostbolt 21.1%, frost nova 8.8%, deep freeze 8.0%).

### Ops: movement log rotated v1 -> v2

`AiPlayerbot.MlDuelMovement.LogFile` now points at **`ml_movement_duel_v2.csv`** for the round-3 farm.

`ml_movement_duel_v1.csv` is M1's frozen `data_tag` (see `artifacts/duel/m1/manifest.json`) but had grown to **8.6 GB** by accumulating M2-era rows behind that name. M2 retrains the ability head only (DEC-042 alternation depth 1), so nothing trains on the movement log now and the two eras have no reason to share a file. v1 is archived to `AzerothCore-server\archive\m1-frozen-20260817.tar.gz` alongside the m1 training windows and gate CSVs.

Caveat recorded rather than edited away: rows after line **34,768,251** of the archived v1 are from a post-sync verification run, not M1 training data. Same for `ml_decisions_duel_m2_r2.csv` after line **2,842,275** - those are valid round-2-config farm rows, just produced post-sync, so round-3 aggregation can keep them.

### Still open

Round-3 farm remains **wired, not launched**. Given round 2 measures -18.8pp on the mage seat, worth deciding before spending the farm time whether round 3 as recipe'd (more win-anchored data on a world the warrior dominates) addresses the starvation or compounds it.

**Comment by gamesh411** [comment]
## Session wrapped - handoff landed, round 3 still not started

Continuation doc: [`docs/ml/research/handoff-2026-08-18-upstream-sync-m2-gate.md`](https://github.com/gamesh411/mod-playerbots/blob/exp/duel-rl-curriculum-postsync/docs/ml/research/handoff-2026-08-18-upstream-sync-m2-gate.md)

### Where the code is

The rebase rewrote `exp/duel-rl-curriculum`, so its pointer on `fork` still sits at the pre-rebase `ce520262`. Nothing is at risk - the full rebased history plus the handoff is on the remote three ways:

| Ref | Points at |
|-----|-----------|
| `pre-upstream-sync-20260816` | `d6c5b180` - the pre-rebase state |
| `post-upstream-sync-20260816` | `589e9bd5` - the rebased state |
| `exp/duel-rl-curriculum-postsync` | `41a177c4` - rebased state + this handoff |

Outstanding is only the branch pointer move on `exp/duel-rl-curriculum`, which needs `--force-with-lease`.

Orchestrator side: `wotlk-playerbots-server` `9fe76a6` pushed - upstream-sync plumbing (fork remote, branch pin, line-ending-safe patches) and the movement log rotation, as two commits.

### Round 3 is wired, not launched

Deliberately left unstarted. The live conf is back on the round-3 farm config (not the tau=0 gate config the measurement runs used), so it is one command away whenever the go/no-go lands. Anti-thrash budget untouched at 2.

**Comment by gamesh411** [comment]
## Resolved - DEC-051: round 3 cancelled, M2 frozen on per-seat peaks (first both-seat gate PASS of the curriculum)

The round-3 go/no-go held open by the last two sessions is resolved by measurement, and the answer made round 3 unnecessary: **both seats already owned a gate-passing artifact**. M2 is frozen as **`stage/m2-ability-coadapt`**.

Module `addc66e0` (freeze commit, tagged), orchestrator `2090792`. The outstanding `exp/duel-rl-curriculum` pointer move from the handoff is done (`ce520262` -> `addc66e0`, `--force-with-lease`).

### The measurement that decided it

The r0-BC and r1-win heads had never been WR-measured - deploys rode offline gates only. Backfilled tau=0 gate legs (identical protocol and baseline as the round-2 gate; >=3.2k matches each; ~40 min per leg) completed the per-round trajectory:

| Head | Warrior WR (need >=47.9%) | Mage WR (need >=56.6%) |
|------|---------------------------|-------------------------|
| r0-BC | 33.0% | **75.2% - PASS (+20.6pp, ~23 sigma)** |
| r1-win | **58.0% - PASS (+12.1pp, ~13 sigma)** | 49.7% |
| r2-win | 46.9% | 35.8% |

Runner control: the new leg runner reproduced the round-2 frost measurement at **35.4%** vs the recorded 35.8% - all four backfilled numbers inherit the round-2 gate's protocol validity.

So the question the go/no-go asked ("does more win-anchored data fix or compound the mage?") had a sharper answer than either option: the loop never helped the mage at all (monotone 75.2 -> 49.7 -> 35.8), the warrior peaked at round 1, and **round 2 was a regression on both seats** that the per-deploy degeneracy gate could not see. Round 3 as wired would have farmed a world where both seats play their worst certified policies.

### Why the trajectory looks like this (DEC-051 in DECISIONS.md)

- **The mage r0 peak is DEC-049 working as designed.** The BC target is stock's relevance argmax *intersected with the castable set*, so the clone plays stock's rotation minus its wasted uncastable picks, and it summons the elemental through the learned head. Cloning a strong teacher through a legality projection beat the teacher by 20.6pp at round 0.
- **Win-only label quality tracks opponent strength, in both directions.** The warrior's +25pp round-1 jump trained on wins earned against the strong r0-BC mage; the mage's round-1 collapse trained on wins earned against the 33% r0 warrior. Round 2 then regressed both: the warrior on easy wins vs the collapsing mage (won_frac 0.718 - junk positives), the mage on thin survivorship-biased wins vs the strong warrior (won_frac 0.242, 428k rows).

### Freeze record (DEC-019)

- Canonical: `artifacts/duel/m2/warrior.pbml` <- `warrior.r1-win.pbml`, `mage.pbml` <- `mage.r0-bc.pbml` (sha-verified).
- Manifest + stage card `docs/ml/curriculum/m2-ability-coadapt.md`; curriculum README status row; tag `stage/m2-ability-coadapt` pushed.
- Deploy gates on record for both frozen heads (degeneracy; mage DEC-050 pet-down floor 57.7% vs floor 25%).
- `duel-m2` replay profile landed (DEC-040 extension) with a certified-era override: M2 is the first stage certified in the DEC-044 world, so its replay keeps `PetReset=1` / `UnglyphedMageShare=0.25` (per-stage exception to the pre-DEC-044 replay default). Verified end to end with `-NoStart`: pins, tau=0, demo CSV, profile verification OK.
- Showcase rows (report-only, same fresh baseline): dec033r3+M1 warrior 44.6% / mage 38.8%; dec033r3+M0 warrior 50.2% / mage 37.6% (M0 rows carry a baseline-mismatch caveat - scripted-movement world vs M1-movement baseline). The statue-era heads sit at stock-or-worse; the frozen M2 pair beats them on both seats.
- Gate CSVs on disk: `ml_decisions_duel_m2gate_{frost,arms}_{r0,r1}.csv`, `..._frost_r2repro.csv`, `ml_decisions_duel_m2show_{arms,frost}_{m1,m0}.csv`.

### Loop-discipline consequence (forward-looking, in DEC-051)

The per-deploy degeneracy gate asks "is any action dead", never "is this policy good" - it passed both round-2 regressions. From M2 on, **a head earns canonical deployment only with a tau=0 WR leg vs the stage baseline** on top of the offline gates. A 40-minute leg after each of rounds 1-2 would have caught both regressions before ~12 h of farm ran on their worlds.

The anti-thrash budget expires unspent; the DEC-044 scaffolding fallback was never triggered (the frozen mage summons in 32.7%+ of tau=0 duels).

### Ops state

- Servers stopped; conf restored to the `duel-farm` state (canonical paths now load the frozen peaks; the `ml_decisions_duel_m2_r3.csv` log name remains wired but the round-3 recipe is retired - nothing launches it).
- The M2 curriculum arc (M0 -> M1 -> M2) is complete: the map's destination stages are all frozen or waived, and M2 is the first with a both-seat uplift PASS.
